### PR TITLE
External tool execution v0.1

### PR DESCRIPTION
**Pull request by HLWeil**
#75 

These changes allow the user to run external executables via the ArcCommander.

At the moment, the tool needs to be placed somewhere in ARC. When running the ArcCommander with an undefined command, the ArcCommander will check whether an executable with the same name as this undefined command can be found somewhere in the ARC directory. 
This executable then gets started. All arguments behind this undefined command and a path argument are given as arguments to this executable.

E.g.: 
A user runs `arc assay geoconvert -a MyAssay` on a windows system
-> ArcCommander doesn't know `geoconvert`
-> ArcCommander will check if it can find a `geoconvert.exe` 
-> ArcCommander will run `geoconvert.exe` `-p TheArcPath` `-a MyAssay`

### EDIT:
After input from @chgarth, naming convention was switched. Instead of searching for a tool named after the first unknown command in the argument chain. Appends all all commands up until this first unknown argument. The example above now is:

A user runs `arc assay geoconvert -a MyAssay` on a windows system
-> ArcCommander doesn't know `geoconvert`
-> ArcCommander will check if it can find a `arc-assay-geoconvert.exe` 
-> ArcCommander will run `arc-assay-geoconvert.exe` `-p TheArcPath` `-a MyAssay`

### CODE REVIEW

**Comment by chgarth** [comment]
Nice work! I would suggest to adopt/extend Git's convention for extension commands:

- If `arc` doesn't know the subcommand xxx (i.e. called as `arc xxx <args...>`), then it tries to execute `arc-xxx <args...>`
- For `arc assay xxx <args...>`, it would try to execute `arc-assay-xxx <args...>`
- No custom search for executable is performed, rather, the extension executable should be found within $PATH/%PATH%.

(If needed, PATH can be extended before invoking the command to allow for custom locations.)

**Comment by HLWeil** [comment]
Switched to new naming convention in febcf5c973f5ecb607283d91b73c6ad64a579c72

See:
```
PS C:\Users\HLWei\source\repos\samplearc_proteomics> arc assay explode -a abc

Could not parse given commands.
Try checking if executable with given argument name "arc-assay-explode"exists
found executable in path:
        "C:\Users\HLWei\source\repos\samplearc_proteomics\arc-assay-explode.exe"
start process C:\Users\HLWei\source\repos\samplearc_proteomics\arc-assay-explode.exe -p C:\Users\HLWei\source\repos\samplearc_proteomics -a abc
```

**Comment by chgarth** [comment]
Nice! Agree that this can be done in a later release, but let me keep gently pushing towards a (IMHO) better solution right away. 

In essence, what your in-ARC executable search is replicating is the general search the OS does for binaries. In doing your own search, you need to account for platform differences (which files are executable? .exe suffix on Windows, no suffix on other platforms but need x-bit set, ...) and are presuming that these executables reside within the ARC, whereas they could really be anywhere. So from where I'm standing, you would get more functionality/portability by doing less (leave the search to the OS). 

A custom search location (e.g. the ARC directory) can be achieved by simply adding that location to the PATH environment variable. If I see this correctly, this is what `ProcessStartInfo.Environment` is for?

Just my 0.02¢.

**Comment by omaus** [comment]
Easiest solution would be, imo, adding the compiled executable file to the release (ArcCommander executable file) and to call it from its unpacked path (which can easily be automatized).

**Comment by HLWeil** [comment]
@omaus, but this way one would not be able to add an executable themselves easily. I'm not quite sure if this will even be a used feature but I think following @chgarth's suggestion might be a good way to achieve a high flexibility. 

Just to get you correctly @chgarth, you would skip the `tryFindExecutablePath` step and directly try `runExecutable` instead, hoping that an executable with that name can be found somewhere in `PATH`?

**Comment by chgarth** [comment]
Yes. If needed, extra directories to search can be added to PATH.

I was thinking along the following lines (apologies for the ugly F#):

```
open System.Diagnostics

let extraDir = "/tmp/fubar"

let psi = 
    ProcessStartInfo(
        UseShellExecute = false,
        FileName = "printit.sh"
    )

let oldpath = System.Environment.GetEnvironmentVariable("PATH")

// temporarily set extended search path
System.Environment.SetEnvironmentVariable("PATH", oldpath + ":" + extraDir)

let p = new Process(StartInfo = psi)

if p.Start() then
    p.WaitForExit()
else
    printf "failed to start"

// restore old search path
System.Environment.SetEnvironmentVariable("PATH", oldpath)
```

One aspect I'm unsure about: On UNIX, PATH contains at ':'-separated list of directories, while on Windows the separator might be ';'?

**Comment by omaus** [comment]
> One aspect I'm unsure about: On UNIX, PATH contains at ':'-separated list of directories, while on Windows the separator might be ';'?

Exactly.

**Comment by omaus** [comment]
The new feature works as well as intended. With that, the question arises if it should be expanded to other type of (external) files since it covers only executable files atm.
We could also cover, e.g., `.cmd`/`.bat` and `.sh` files.

@chgarth 
@HLWeil

**Comment by chgarth** [comment]
This looks very cool! No reservations about merging.

Minor question: Is it possible to simply call `arc-assay-explode` instead of `arc-assay-explode.exe`? IIRC this should address the `.cmd`/`.bat` use case in Windows. For Unix, this should be a non-problem, since scripts etc. do not have to carry an extension.

**Comment by chgarth** [comment]
> But this does _not_ apply to `.cmd` and `.bat` (`.sh` in Linus similarly), they must be called with the extension (in your example: `arc assay explode.cmd`) (funnily, this is only the case with the .NET runtime and not with CMD and PowerShell)

If this is specific to .NET, would it be possible to wrap execution, i.e. "cmd /c arc-assay-explode <...>"? This would then revert to CMD.exe's executable search.

(For Linux/.sh, I don't think it's a problem since .sh is a convention, rather than a requirement, for execution. So you can have e.g. Bash scripts without the .sh extension, but the x-bit set and a shebang.)

**Comment by omaus** [comment]
> If this is specific to .NET, would it be possible to wrap execution, i.e. call "cmd /c arc-assay-explode <...>"? This would then revert to CMD.exe's executable search.

Yes, but this would result in a new CMD window (already tested). It's the same like setting the `UseShellExecute` field of the `ProcessStartInfo` object to `true`.
Because of this I would suggest trying all possible extensions until one succeeds. If no possibility works, return an according message. What do you think? @chgarth

**Comment by chgarth** [comment]
> So this is a possible solution, @chgarth.

Looks good to me, thanks for indulging me! As the final piece of the puzzle, could you check whether `cmd /C` still allows to feed/capture stdin/stdout/stderr?

**Comment by omaus** [comment]
> As the final piece of the puzzle, could you check whether cmd /C still allows to feed/capture stdin/stdout/stderr?

No, unfortunately this is not the case. The ArcCommander is not able to capture the output of an app called this way.
Feed arguments into the external tool only works via parsing it as an argument (hardcoded). Feed the tool while it runs (e.g. "do you want to proceed [y/n]?") from the user's side works when coded into the external tool.
How big of a problem is this?

**Comment by chgarth** [comment]
> How big of a problem is this?

For the intended use case (extension executables), it IMHO shouldn't be a problem. So I think we have a solution for the moment.

**Review by HLWeil — COMMENTED** [review]
Nice improvements! If possible, a cleanup touch would be nice and then it's good to go 👍

File: src/ArcCommander/Program.fs [hunk 1]
```diff
@@ -227,39 +265,95 @@ let main argv =
...
+            with
+            | e2 -> 
+                // Incorrectly parsed arguments will be threaded into external executable tool handler
+                // Here for the first unknown argument in the argument chain, an executable of the same name will be called
+                // If this tool exists, try executing it with all the following arguments
+                log.Info("Could not parse given commands.")
+                match tryGetUnknownArguments parser argv with
+                | Some (executableNameArgs, args) ->
+                    let executableName = (makeExecutableName executableNameArgs)
+                    if executableName = "arc-ducklings" then StudyAPI.Protocols.playAllMyLittleDucklings()
```
[the remark this answers is not in the document]
Comment (HLWeil) [reply src/ArcCommander/Program.fs:291]: Somehow the edit went missing, I saw you already created the `fs`

**Comment by HLWeil** [comment]
That's a go from my side too! 👍

**Pull request merged**